File: C++/Stacks & Queues/11_evaluation_of_postfix_expression.cpp
/*
Idea:
using stack, inserting till numbers are found ,if any operator is encountered,
performing operations on top 2 elements and pushing the answer on stack, performing till stack has only 1 element.
*/

//{ Driver Code Starts
// C++ program to evaluate value of a postfix expression

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution
{
public:
    //Function to evaluate a postfix expression.
    int evaluatePostfix(const string& expression)
    {
        stack<int> stack;

        for (char ch : expression)
        {
            if (isdigit(ch))
            {
                // If the character is a digit, convert it to an integer and push onto the stack
                stack.push(ch - '0');
            }
            else
            {
                // If the character is an operator, pop the top two elements, perform the operation, and push the result back
                int operand2 = stack.top();
                stack.pop();
                int operand1 = stack.top();
                stack.pop();

                switch (ch)
                {
                case '+':
                    stack.push(operand1 + operand2);
                    break;
                case '-':
                    stack.push(operand1 - operand2);
                    break;
                case '*':
                    stack.push(operand1 * operand2);
                    break;
                case '/':
                    stack.push(operand1 / operand2);
                    break;
                default:
                    break;
                }
            }
        }
        // The result should be the only element left in the stack
        return stack.top();
    }
};

//{ Driver Code Starts.

// Driver program to test above functions
int main()
{
    int t;
    cin >> t;
    cin.ignore(INT_MAX, '\n');
    while (t--)
    {
        string S;
        cin >> S;
        Solution obj;

        cout << obj.evaluatePostfix(S) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Python/Stacks & Queues/11_evaluation_of_postfix_expression.py
"""
Idea:
using stack, inserting till numbers are found ,if any operator is encountered,
performing operations on top 2 elements and pushing the answer on stack, performing till stack has only 1 element.

"""


# User function Template for python3

class Solution:

    # Function to evaluate a postfix expression.
    def evaluatePostfix(self, expression):
        stack = []

        for char in expression:
            if char.isdigit():
                stack.append(int(char))
            else:
                operand2 = stack.pop()
                operand1 = stack.pop()

                if char == '+':
                    stack.append(operand1 + operand2)
                elif char == '-':
                    stack.append(operand1 - operand2)
                elif char == '*':
                    stack.append(operand1 * operand2)
                elif char == '/':
                    stack.append(operand1 // operand2)

        # The result should be the only element left in the stack
        return stack[0]


# code here


# {
# Driver Code Starts
# Initial Template for Python 3

import atexit
import io
import sys


_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        S = input()
        obj = Solution()
        print('{0:g}'.format(float(obj.evaluatePostfix(S))))
# } Driver Code Ends